test_apps/smokeTests/gridTools.cpp

```diff
@@ -0,0 +1,436 @@
+    float yIncrement = axis==Y ? (maxVal - minVal) / (dims[Y]-1) : 0;
+    float zIncrement = axis==Z ? (maxVal - minVal) / (dims[Z]-1) : 0;
+
+    float value = minVal;
+    for ( size_t k=0; k<z; k++ ) {
+        value = axis==Y ? minVal : value;      // reset value if we're ramping on Y
+        for ( size_t j=0; j<y; j++ ) {
+            value = axis==X ? minVal : value;  // reset value if we're ramping on X
+            for ( size_t i=0; i<x; i++ ) {
+                grid->SetValueIJK( i, j, k, value );
+                value += xIncrement;
+            }
+            value += yIncrement;
+        }
+        value += zIncrement;
+    }
+}
+
+// This function iterates across all nodes in a grid; making comparisons between the
+// data values returned by the functions GetValueAtIndex() and GetValue().
+int CompareIndexToCoords( 
+    VAPoR::Grid* grid, 
+    double &rms,                // Root Mean Square error
+    size_t &numMissingValues,   // Counter for receiving MissingValue upon query
+    size_t &disagreements      // Counter for when AccessIJK() and GetValue() disagree
+) {
+    rms   = 0.f;
+    disagreements  = 0;
+    numMissingValues = 0;
+
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+ 
+    double peak  = 0.f;
+    double sum     = 0;
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                size_t indices[3] = { i, j, k };
+                //float trueValue = grid.AccessIJK( i, j, k );
+                double trueValue = grid->GetValueAtIndex( {i,j,k} );
+                
+                double coords[3];
+                grid->GetUserCoordinates( indices, coords );
+                float sampleValue = grid->GetValue( coords );
+                
+                if (sampleValue == grid->GetMissingValue()) {
+                    numMissingValues++;
+                    continue;
+                }
+
+                double error = abs( sampleValue - trueValue );
+
+                if (error != 0) {
+                    disagreements++;
+                }
+
+                if ( error > peak ) 
+                    peak = error;
+                sum += error*error;
+
+            }
+        }
+    }
+
+    rms = sqrt( sum / (x*y*z) );
+
+    if ( rms != 0 || disagreements > 0 )
+        return 1;
+    return 0;
+}
+
+size_t TestNodeIterator( const Grid *g, int& count, double& time ) {
+    Grid::ConstNodeIterator itr;
+    Grid::ConstNodeIterator enditr = g->ConstNodeEnd();
+
+    double t0 = Wasp::GetTime();
+    itr = g->ConstNodeBegin();
+
+    for ( ; itr!=enditr; ++itr) {
+        count++;
+    }
+
+    size_t expectedCount = 1;
+    std::vector< size_t > dims = g->GetDimensions();
+    for ( auto dim : dims )
+        expectedCount *= dim;
+
+    time = Wasp::GetTime() - t0;
+
+    return expectedCount;
+}
+
```

Comment:
It would be good to have some additional iterator tests. We have three iterator types:

Grid::Iterator // dereferences data values
Grid::ConstCoordItr // dereferences node coordinates
Grid::ConstNodeIterator // dereferences node indices

A simple test for each would be to verify that the dereferenced values are the same as what you get with explicit access using the accessors.

Reply:
@clyne - Done.  I believe there are bugs in the ConstCoordIterator, for the Regular and Layered grids.  Corresponding with more details over email.